src/typechecker.erl

```diff
@@ -3388,11 +3388,60 @@ check_guard_call(is_reference, [{var, _, Var}]) -> #{Var => type(reference)};
 check_guard_call(is_tuple, [{var, _, Var}]) -> #{Var => type(tuple, any)};
 check_guard_call(_Fun, _Vars) -> #{}.
 
+%% Derive a type from a comparison against a literal value
+%% using Erlang's term order:
+%%
+%%     number < atom < reference < fun < port < pid <
+%%            < tuple < map < nil < list < bitstring
+%%
+%% Note: This function may return non-standard ranges like 10..pos_inf
+type_comp_op('=<', I) when is_integer(I) ->
+    IntTypes = gradualizer_int:int_range_to_types({neg_inf, I}),
+    type(union, [type(float) | IntTypes]);
+type_comp_op('<', I) when is_integer(I) ->
+    type_comp_op('=<', I - 1);
+type_comp_op('>=', I) when is_integer(I) ->
+    IntTypes = gradualizer_int:int_range_to_types({I, pos_inf}),
+    GtTypes = [type(atom), type(reference), type('fun'), type(port), type(pid),
+               type(tuple, any), type(map, any), type(list), type(bitstring)],
+    type(union, IntTypes ++ GtTypes);
+type_comp_op('>', I) when is_integer(I) ->
+    type_comp_op('>=', I + 1);
+type_comp_op('==', I) when is_integer(I) ->
+    type(union, [type(float), {integer, erl_anno:new(0), I}]);
+type_comp_op('=:=', I) when is_integer(I) ->
+    {integer, erl_anno:new(0), I};
+type_comp_op(Op, A) when Op =:= '==' orelse Op =:= '=:=',
+                         is_atom(A) ->
+    {atom, erl_anno:new(0), A};
+type_comp_op(Neq, I) when Neq =:= '/=' orelse Neq =:= '/=',
+                          is_integer(I) ->
+    {type, _, union, U1} = type_comp_op('<', I),
+    {type, _, union, U2} = type_comp_op('>', I),
+    type(union, U1 ++ U2);
```

Comment:
Do you need to normalize the union here?

Reply:
It should already contain only normalized types here, or at least it would be better that way. Now, that I think about it, `type('fun')` , `type(list)` and `type(bitstring)` should return the normalized forms `fun((...) -> any())`, `[any()]` and `<<_:0, _:_*1>>`. I think specific clauses in `type/1` should do that.